src/Processors/Formats/Impl/ORCBlockInputFormat.cpp

```diff
@@ -51,6 +61,26 @@ void ORCBlockInputFormat::resetParser()
     IInputFormat::resetParser();
 
     file_reader.reset();
+    include_indices.clear();
+    prepareReader();
+}
+
+void ORCBlockInputFormat::prepareReader()
+{
+    THROW_ARROW_NOT_OK(arrow::adapters::orc::ORCFileReader::Open(asArrowFile(in), arrow::default_memory_pool(), &file_reader));
+    stripe_total = file_reader->NumberOfStripes();
+    stripe_current = 0;
+
+    std::shared_ptr<arrow::Schema> schema;
+    THROW_ARROW_NOT_OK(file_reader->ReadSchema(&schema));
+
+    for (int i = 0; i < schema->num_fields(); ++i)
+    {
+        if (getPort().getHeader().has(schema->field(i)->name()))
+        {
+            include_indices.push_back(i);
```

Comment:
So we have two sets: a set of column actually stored in file and a set of columns from the header. With this code a header could contains "extra" columns, which are not in the file. For example: file schema is `a UInt64` and header is `a UInt64, b Uint64, c UInt64`. What will be in this case? I think we have to throw an exception. But remember that in situation where a set of columns from header is a subset of columns from file is Ok.

Reply:
@nikitamikhaylov 

If header set has "extra" columns than file set it will throw an exception as below:

```
ubuntu01 :) select ClientEventTime, MobilePhoneModel, ClientIP6, NAME, NAME1, NAME2  from 00163_column_oriented;

SELECT
    ClientEventTime,
    MobilePhoneModel,
    ClientIP6,
    NAME,
    NAME1,
    NAME2
FROM `00163_column_oriented`

Query id: ffd322a5-ad16-4be6-91b3-0395e5d6273e


0 rows in set. Elapsed: 0.005 sec.

Received exception from server (version 21.5.1):
Code: 47. DB::Exception: Received from localhost:9000. DB::Exception: Missing columns: 'NAME1' 'NAME' 'NAME2' while processing query: 'SELECT ClientEventTime, MobilePhoneModel, ClientIP6, NAME, NAME1, NAME2 FROM `00163_column_oriented`', required columns: 'ClientEventTime' 'ClientIP6' 'NAME2' 'MobilePhoneModel' 'NAME' 'NAME1', maybe you meant: ['ClientEventTime','ClientIP6','MobilePhoneModel'].
```
So the exception has been throw by other place.

Reply:
Ok.